Fixing some iterator issues

vakokako opened the pull request:
This merge request fixes some issues with iterators.

- [x] `xfunctor_view::crbegin` was returning non const iterator
- [x] `xiterator_adaptor::data` was returning iterator, not a pointer
- [x] `xt::reshape_view().rbegin()` didn't compile

### Fixing multidimensional index of end `xiterator`
- [x] multidimensional index of `xiterator` for `end()` iterator would be always set to `shape()`, which made `*rbegin()` be out of bounds access. I changed the index of `end()` to be `shape[i] - 1` for all dimensions, except the last one in row-major case, and except the first one in column-major case

```cpp
xarray<int> a = { {{0, 1, 2, 3},
                   {4, 5, 6, 7},
                   {8, 9, 10, 11}},
                  {{12, 13, 14, 15},
                   {16, 17, 18, 19},
                   {20, 21, 22, 23}} };

auto view = dynamic_view(a, { 1, keep(0, 2), range(1, 4) });
// shape: {1, 2, 3}

// old version
view.end(); // index {1, 2, 3}
std::prev(view.end()); // index {1, 2, 2}
// out of bounds access
*view.rbegin(); // index {1, 2, 2}

// new version
view.end(); // index {0, 1, 3}
std::prev(view.end()); // index {0, 1, 2}
// correct access
*view.rbegin(); // index {0, 1, 2}
```

tdegeus commented on include/xtensor/xbuffer_adaptor.hpp:
```diff
@@ -866,13 +921,13 @@ namespace xt
     inline auto xbuffer_adaptor_base<D>::operator[](size_type i) -> reference
     {
-        return derived_cast().data()[static_cast<index_type>(i)];
+        return begin()[static_cast<index_type>(i)];
```
Could this be `data` two?

vakokako replied:
yep

tdegeus commented on include/xtensor/xbuffer_adaptor.hpp:
```diff
@@ -1188,13 +1255,25 @@ namespace xt
-    inline auto xiterator_owner_adaptor<C, IG>:: data() noexcept -> iterator
+    inline auto xiterator_owner_adaptor<C, IG>::data() noexcept -> pointer
+    {
+        return &(*m_it);
```
Sorry, maybe I need coffee, but is this not either useless, or illegal. Useless because it will return the same memory address. Or, illegal, because it might point to a temporary?

Also: Will this break things for users?

JohanMabille replied:
This is indeed equivalent to `m_it` when dereferencing `m_it` returns an lvalue, and illgal when it return an rvalue, so this should be reverted to the previous implementation.

vakokako replied:
this was done to make data() consistently return `pointer` instead of `iterator` in this case, but it's indeed faulty and not needed here.

tdegeus reviewed (commented):
Thanks. I have not looked at everything yet, but I do have some questions already

vakokako commented:
Hey, are there any updates on including this in xtensor?

emmenlau commented:
> Sorry for missing this, I have been swamped with a lot of other tasks. There is an issue in this PR that I want to comment on since it is not trivial. I'll try to do it next week.

Hey @JohanMabille , you said there are some issues with this PR, can you tell us a hint where to look?

JohanMabille commented on include/xtensor/xbuffer_adaptor.hpp:
```diff
@@ -1092,13 +1147,25 @@ namespace xt
-    inline auto xiterator_adaptor<I, CI>::data() noexcept -> iterator
+    inline auto xiterator_adaptor<I, CI>::data() noexcept -> pointer
+    {
+        return &(*m_it);
```
This is incorrect for iterator returning rvalues when dereferenced. This should be reverted to the previous implementation, except if this was leading to a bug. In that case, `xtl::xclosure_wrapper / xtl::xclosure_pointer` should be used instead (but I'm really not sure we need them here).

JohanMabille commented on include/xtensor/xstrided_view.hpp:
```diff
@@ -219,9 +221,18 @@ namespace xt
+        const_reverse_linear_iterator linear_rbegin() const;
+        const_reverse_linear_iterator linear_rend() const;
+        const_reverse_linear_iterator linear_crbegin() const;
+        const_reverse_linear_iterator linear_crend() const;
```
Godd catch, thanks for adding them ;)

JohanMabille reviewed (changes requested):
Again sorry for the very late answer. I think I was confused with another fix on the iterators we tried in the past, and TBH I don't remember the detail. Anyway, this fix regarding the last iterator index is the way to go.

I'm not sure we need the changes regarding the buffer adaptor's API:
- the `begin_impl/end_impl` would be required if it had a diffrerent implementation from the `data` method, and in that case they would be private. Since `data` already do what we need, no need to add public API / additional code to maintain IMO ;)

Would it be possible to remove these changes from this PR? If you think they are still relevant, we can discuss them in a dedicagted PR. I'll be watchful to the changes to this PR and review them ASAP

vakokako commented:
@JohanMabille I updated the merge request, removed the changes to buffer adaptor's API. There is a failing 'pretty format json' pre-commit check, I'm not sure what does that mean and how to deal with that?

JohanMabille commented:
> [... ] There is a failing 'pretty format json' pre-commit check, I'm not sure what does that mean and how to deal with that?

I think running `pre-commit run --all-files` and push the changes should fix the issue. This failure is probably due to an update of pre-commit since your PR does not change anything regarding the notebooks.

JohanMabille commented:
Actually this also requires updating the config, let's merge this one and I'll fix the build on master directly.

vakokako commented:
Thanks!
